src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -905,6 +963,138 @@ private void EmitArray<T>(Type arrType, IEnumerable<T> items, Action<T, int> ite
+                    //TODO remove IList from this if stmt when we figure out function call return wrapping behavior.
+                    //this is still a problem for BuiltIn Dictionaries that are wrapped in an IList.
+                    if (arrayT == null || arrayT == typeof(IList) || arrayT == typeof(object))
+                    {
+                        return (false, null);
+                    }
+                }
+                else
+                {
+                    return (false, null);
+                }
+            }
+          
+            //emit load array to stack.
+            var t = DfsTraverse(array);
+
+            if (t == null)
+            {
+                return (false, null);
+            }
+            else if (typeof(IDictionary).IsAssignableFrom(t))
+            {
+                if (t.IsGenericType)
+                {
+                    TryEmitIndexingForDictionary(fcn.FormalArguments[0], fcn.FormalArguments[1], t);
+                    return (true, t.GenericTypeArguments[0]);
+                }
+                else
+                {
+                    TryEmitIndexingForDictionary(fcn.FormalArguments[0], fcn.FormalArguments[1], t);
+                    return (true, typeof(object));
+                }
+            }
+            //builtin DS dict is a wrapper
+            else if (typeof(DesignScript.Builtin.Dictionary).IsAssignableFrom(t))
+            {
+                //TODO
+                //emit function call for ValueAtKey or fallback to replication.
+            }
+
+            else if (t.IsArray)
+            {
+                TryEmitIndexingForArray(fcn.FormalArguments[0], fcn.FormalArguments[1], t.GetElementType());
+                return (true, t.GetElementType());
+            }
+            // TODO we may want to bail for IList currently and let 
+            // replication handle it as Ilist is usually a replicated output, and is nested.
+            //Today either we end up emitting IList index opcodes incorrectly for all func return vals,
+            //or we'll call the wrong overload until overload matching is fixed.
+            else if (t == typeof(IList))
+            {
+                if (t.IsGenericType)
+                {
+                    TryEmitIndexingForIList(fcn.FormalArguments[0], fcn.FormalArguments[1], t, t.GenericTypeArguments[0]);
+                    return (true, t.GenericTypeArguments[0]);
+                }
+                else
+                {
+                    TryEmitIndexingForIList(fcn.FormalArguments[0], fcn.FormalArguments[1], t, typeof(object));
+                    return (true, typeof(object));
+                }
+            }
+            EmitILComment("NOT ENOUGH TYPE INFO TO EMIT INDEXING");
+            return (false, null);
+        }
+
+
+        private bool TryEmitIndexingForIList(AssociativeNode array, AssociativeNode index, Type collectionType, Type listElementType)
```

Comment:
Looks like these methods always will succeed now and return `true`. Should we drop the `Try` part and make them void?

Reply:
done